### ISSUE TITLE
Show prototypes in documentSymbols

### ISSUE THREAD

**Issue by sergei-dyshel** [MaskRay/ccls#122]
When opening a header file I'd like document symbols to show also function prototypes/declarations. `clangd` does this BTW.

Tried to do it on my own and replaced `sym.role & Role::Definition` by `sym.role & Role::Definition || sym.role & Role::Declaration` in couple of places in `src/messages/textDocument_documentSymbol.cc` but it didn't work. What am I missing?

**Comment by MaskRay** [comment]
If your client supports hierarchical document symbol https://github.com/MaskRay/ccls/wiki/Status-table
```cpp
struct A {
    void foo(); // definition in .cc file
    void bar() {}
};
```

Member functions have entries in hierarchy document symbol.

Can you give some concrete example?

**Comment by sergei-dyshel** [comment]
For example here is the outline (uses documentSymbol) of `ccls`'s `pipeline.hh` in vscode:
![image](https://user-images.githubusercontent.com/6907277/48647459-b3a49580-e9f4-11e8-9412-75658f2c36c6.png)

As you can see, class member prototypes are shown (eg., `Clear` or `Stamp`) but non-class function prototypes are not shown (such as `pipeline::Index`).

**Comment by sergei-dyshel** [comment]
@MaskRay 
Now, after updating to the last version, I have a reverse problem - method definitions in source files are not show. For example for `src/messages/textDocument_hover.cc` I'm getting empty outline in vscode.

**Comment by MaskRay** [comment]
I checked `textDocument_hover.cc` and also a simple example:
```cpp
// a.cc
#include "a.h"
void A::foo() {}
```

Both work.

**Comment by sergei-dyshel** [comment]
Does this work for you only in Emacs or you check with vscode too?

Anyway, I think I found the root cause. According to LSP spec, `selectionRange` must be contained in `range`:
```
/**
	 * The range enclosing this symbol not including leading/trailing whitespace but everything else
	 * like comments. This information is typically used to determine if the clients cursor is
	 * inside the symbol to reveal in the symbol in the UI.
	 */
	range: Range;

	/**
	 * The range that should be selected and revealed when this symbol is being picked, e.g the name of a function.
	 * Must be contained by the `range`.
	 */
	selectionRange: Range;
```

However, in response for documentSymbol by ccls I've found non-complying entries, such as:
```
 {
                "name": "Reflect",
                "detail": "ccls::(anon ns)::Reflect",
                "kind": 12,
                "range": {
                    "start": {
                        "line": 41,
                        "character": 0
                    },
                    "end": {
                        "line": 41,
                        "character": 19
                    }
                },
                "selectionRange": {
                    "start": {
                        "line": 41,
                        "character": 0
                    },
                    "end": {
                        "line": 41,
                        "character": 43
                    }
                },
```

I think that wherever vscode receives such an entry, it just throws away the whole response. In order to check my hypothesis I've commented out the following lines in `textDocument_document.cc`:
```
        ds->selectionRange = *range;
        ds->range = ds->selectionRange;
        // if (sym.extent.Valid())
        //   if (auto range1 = GetLsRange(wfile, sym.extent))
        //     ds->range = *range1;
```

And everything worked. For now staying with this workaround and waiting for proper fix :smile:

**Comment by sergei-dyshel** [comment]
Yes, I remember seeing more than 1 `Reflect` symbol in outline despite only one being in the source. Frankly, I think the filtering criteria in `textDocument_document.cc` need more polishing. For example, in `textDocument_hover.cc` I see those `d`, `m` `def` etc., which are probably function arguments or local variables and IMHO shouldn't be part of outline.
